pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -128,345 +167,552 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+
+	// set the binding reference to the secret
+	// https://github.com/external-secrets/external-secrets/pull/2263
+	es.Status.Binding = v1.LocalObjectReference{Name: newSecret.Name}
+
+	r.recorder.Event(es, v1.EventTypeNormal, esv1beta1.ReasonCreated, "Created Secret")
+	return nil
 }
 
-func (r *Reconciler) patchSecret(ctx context.Context, secret *v1.Secret, mutationFunc func() error, es *esv1beta1.ExternalSecret) error {
+func (r *Reconciler) updateSecret(ctx context.Context, existingSecret *v1.Secret, mutationFunc func(secret *v1.Secret) error, es *esv1beta1.ExternalSecret, secretName string) error {
 	fqdn := fmt.Sprintf(fieldOwnerTemplate, es.Name)
-	err := r.Client.Get(ctx, client.ObjectKeyFromObject(secret), secret.DeepCopy())
-	if apierrors.IsNotFound(err) {
-		return fmt.Errorf(errPolicyMergeNotFound, secret.Name)
-	}
-	if err != nil {
-		return fmt.Errorf(errPolicyMergeGetSecret, secret.Name, err)
-	}
 
-	existing := secret.DeepCopyObject()
-	if err = mutationFunc(); err != nil {
-		return fmt.Errorf(errPolicyMergeMutate, secret.Name, err)
+	// fail if the secret does not exist
+	// this should never happen because we check this before calling this function
+	if existingSecret.UID == "" {
+		return fmt.Errorf(errUpdateNotFound, secretName)
 	}
 
-	// GVK is missing in the Secret, see:
-	// https://github.com/kubernetes-sigs/controller-runtime/issues/526
-	// https://github.com/kubernetes-sigs/controller-runtime/issues/1517
-	// https://github.com/kubernetes/kubernetes/issues/80609
-	// we need to manually set it before doing a Patch() as it depends on the GVK
-	gvks, unversioned, err := r.Scheme.ObjectKinds(secret)
-	if err != nil {
-		return err
-	}
-	if !unversioned && len(gvks) == 1 {
-		secret.SetGroupVersionKind(gvks[0])
+	// set the binding reference to the secret
+	// https://github.com/external-secrets/external-secrets/pull/2263
+	es.Status.Binding = v1.LocalObjectReference{Name: secretName}
+
+	// mutate a copy of the existing secret with the mutation function
+	updatedSecret := existingSecret.DeepCopy()
+	if err := mutationFunc(updatedSecret); err != nil {
+		return fmt.Errorf(errMutate, updatedSecret.Name, err)
 	}
 
-	if equality.Semantic.DeepEqual(existing, secret) {
+	// if the secret does not need to be updated, return early
+	if equality.Semantic.DeepEqual(existingSecret, updatedSecret) {
 		return nil
 	}
-	// Cleaning up Managed fields manually as to keep patch coherence
-	secret.ObjectMeta.ManagedFields = nil
-	// we're not able to resolve conflicts so we force ownership
-	// see: https://kubernetes.io/docs/reference/using-api/server-side-apply/#using-server-side-apply-in-a-controller
-	if err := r.Client.Patch(ctx, secret, client.Apply, client.FieldOwner(fqdn), client.ForceOwnership); err != nil {
-		return fmt.Errorf(errPolicyMergePatch, secret.Name, err)
+
+	// if the existing secret is immutable, we can only update the object metadata
+	if existingSecret.Immutable != nil && *existingSecret.Immutable {
```

Comment:
```suggestion
	if ptr.Deref(existingSecret.Immutable, false) {
```

Reply:
Done in https://github.com/external-secrets/external-secrets/pull/4086/commits/d40d76960900a1be45e13cb3110b06bc6cb9e84f